### PR TITLE
DINO Model addition for Ultralytics Library

### PR DESCRIPTION
**Pull request by kaushal07wick**
This is a PR regarding adding the DINO model to ultralytics library, it is in the first phase now, some modifications are still left.

will update them very soon. This PR will help to keep track of the changes and also get suggestions from the maintainers @AyushExel .

I have read the CLA Document and I sign the CLA
-->

### Summary
📝🚀🎨

This pull request adds several modules to the `ultralytics/vit/dino` directory, which implement the DINO self-supervised learning method for vision transformers and its evaluation on various tasks and datasets. It also adds a new module `visualize_attention.py` to display the self-attention maps of a DINO model on an input image. The pull request is part of an effort to integrate vision transformers into the ultralytics framework.

> _`DINO` learns to see_
> _Vision transformers abound_
> _Winter of self-supervision_

### Walkthrough
*  Add license headers and import statements for nine modules related to vision transformers and DINO ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a6389d4aa3f5303597ffb4201e118066fcf848e6744930cca28c30fd8d5d717fR1-R300), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-349118a7ca36ca9b963109ee8e099f3d133e28672dfb1bbf414b95a3f280b57eL1-R200), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-5549e1fcf1308147fb789bb9937869fcb767930d1b78f72f39368e294767b6d0L1-R241), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-0cf52e719965c8ed3e3f9d8bba07508407a16e4da24644c3f18bd406ea307bccL1-R280), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-464cf863a8a2bce3c959197796b492003d5cc982a8fe6d7eb5cc2271e1902147L1-R291), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a7311be542da428340e30b4e104758c4194904208244dfe5af69f905f380ea28L1-R150), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-885c81b00c9d68787a0a453719dfb8d79cd71801232b0e4c8dcb4cc345abf0eeL1-R131), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a437aa4b375e72835df5d3b54087aacca06367327824d2a9cd057e88f455beb2L1-R377), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-4995441161ec224128a5e09da94551a040fb68f2e7d30d0eb188e0bb9f11d785L1-R287))
* Define classes and functions for the Vision Transformer and DINO models in `vision_transformer.py` ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-4995441161ec224128a5e09da94551a040fb68f2e7d30d0eb188e0bb9f11d785L1-R287))
* Implement scripts to evaluate DINO on various tasks and datasets, such as copy detection, image retrieval, k-NN classification, linear classification, and video object segmentation, in `eval_copy_detection.py`, `eval_image_retrieval.py`, `eval_knn.py`, `eval_linear.py`, and `eval_video_segmentation.py`, respectively ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a6389d4aa3f5303597ffb4201e118066fcf848e6744930cca28c30fd8d5d717fR1-R300), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-349118a7ca36ca9b963109ee8e099f3d133e28672dfb1bbf414b95a3f280b57eL1-R200), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-5549e1fcf1308147fb789bb9937869fcb767930d1b78f72f39368e294767b6d0L1-R241), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-0cf52e719965c8ed3e3f9d8bba07508407a16e4da24644c3f18bd406ea307bccL1-R280), [link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-464cf863a8a2bce3c959197796b492003d5cc982a8fe6d7eb5cc2271e1902147L1-R291))
* Provide functions to load the DINO models from torch.hub in `hubconf.py` ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a7311be542da428340e30b4e104758c4194904208244dfe5af69f905f380ea28L1-R150))
* Support multinode training with submitit in `run_with_submitit.py` ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-885c81b00c9d68787a0a453719dfb8d79cd71801232b0e4c8dcb4cc345abf0eeL1-R131))
* Create self-attention videos from DINO models in `video_generation.py` ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-a437aa4b375e72835df5d3b54087aacca06367327824d2a9cd057e88f455beb2L1-R377))
* Add a new file `visualize_attention.py` to visualize the self-attention maps of a vision transformer model trained with DINO, based on the original file from the dino repository, but with some modifications and additional features ([link](https://github.com/ultralytics/ultralytics/pull/3347/files?diff=unified&w=0#diff-d70bffd9db7a209ea75368ba56186dcce2fbf4aaf5909d1baa541ffbb094333fL1-R212))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces various enhancements and new modules to the Ultralytics DINO repository.

### 📊 Key Changes
- **New DINO Main Python Script**: A streamlined entry point for DINO training with updated Apache License.
- **Copy Detection Evaluation**: Methodology for evaluating copy detection performance on the Copydays dataset.
- **Image Retrieval Evaluation**: Mechanism to assess image retrieval accuracy on revisited Paris and Oxford datasets.
- **k-NN Evaluation**: A script for evaluating model performance using k-NN on ImageNet.
- **Linear Evaluation**: Supports evaluating linear classifiability on frozen features.
- **Video Segmentation Evaluation**: Introduces functionality for video object segmentation assessment on DAVIS 2017.
- **Hub Configuration for Model Access**: Facilitates access to pre-trained DINO models via Torch Hub.
- **Submitit Integration**: Allows for multi-node training using the Submitit library.
- **Utility Functions**: A compilation of helper functions such as Gaussian Blur, model initialization, and data augmentation.
- **Video Generation Utility**: Provides tools for generating videos from images with focus on self-attention visualization.
- **Vision Transformer**: Core implementation of the Vision Transformer model compatible with DINO.
- **Attention Visualization**: Tooling for visualizing self-attention maps generated by the Vision Transformer model.

### 🎯 Purpose & Impact
- **Enhanced Functionality**: These additions enrich the DINO repository, offering more extensive evaluation capabilities and model usage.
- **Better Usability and Accessibility**: The provided scripts and utilities make it easier for both researchers and practitioners to experiment with DINO.
- **Facilitates Research and Development**: The extended toolkit supports a broader range of tasks and evaluations, promoting further advancements in self-supervised learning and beyond.

### CODE REVIEW

**Comment by AyushExel** [comment]
Let me know when this is ready for review

**Comment by kaushal07wick** [comment]
I have read the CLA Document and I sign the CLA

**Comment by kaushal07wick** [comment]
> Let me know when this is ready for review

Hi, you can review it now and see if there any initial bugs.

**Comment by glenn-jocher** [comment]
@kaushal07wick hi there! Thank you for submitting your PR for review. I'll be happy to take a look and provide feedback. It would be helpful if you could provide some context and details about the changes you made and the modifications that are still pending. This will allow me to better understand the scope of the PR and provide more targeted feedback. Once I have the necessary information, I'll dive into the code and look for any initial bugs or issues. Let's make sure everything is in good shape before moving forward. Thanks again for your contribution!

**Comment by kaushal07wick** [comment]
> @kaushal07wick hi there! Thank you for submitting your PR for review. I'll be happy to take a look and provide feedback. It would be helpful if you could provide some context and details about the changes you made and the modifications that are still pending. This will allow me to better understand the scope of the PR and provide more targeted feedback. Once I have the necessary information, I'll dive into the code and look for any initial bugs or issues. Let's make sure everything is in good shape before moving forward. Thanks again for your contribution!

I have reproduced DINO from the main github code, and tried to assemble it properly into modules.
I am just confused on what to do next so that it can become a part of ultralytics library, like we can call it using ultralytics library.

I would be greatful if you could provide me tips on how to test this code according to ultralytics standard.

thanks

**Comment by AyushExel** [comment]
@kaushal07wick I think you might need to refine this a bit. The code you've copied from the DINO repo needs to be re-written with the Ultralytics package in mind. The license header in each file also needs to be gotten rid of. We don't use argparser as this is a pip package but I see a lot of them in the code files. So get rid of them also. That should significantly bring down the number of lines changed in the PR. After that try to get the model inference running, that's the most important part.

**Comment by kaushal07wick** [comment]
> @kaushal07wick I think you might need to refine this a bit. The code you've copied from the DINO repo needs to be re-written with the Ultralytics package in mind. The license header in each file also needs to be gotten rid of. We don't use argparser as this is a pip package but I see a lot of them in the code files. So get rid of them also. That should significantly bring down the number of lines changed in the PR. After that try to get the model inference running, that's the most important part.

Hi @AyushExel so instead of argparse, what should we use.

**Comment by kaushal07wick** [comment]
So @AyushExel , the DINO directory should have a model.py , predict.py an val.py file right.
I have to use functionalities from ultralyitcs package, to make in native to ultralytics right.
So how should be parse the arguments, if we are not using argparse library.

if would be great if you could point me to any resource explaining how we should make this changes, to make any model ultralytics native.

**Comment by glenn-jocher** [comment]
Hi @kaushal07wick, to make the model compatible with the Ultralytics package, you wouldn't need to use the argparse library for argument parsing. Instead, you can utilize the configurations and arguments provided by the Ultralytics package itself.

Ultralytics provides its own way of handling configuration and arguments specifically tailored for their framework. You can refer to the Ultralytics documentation or their GitHub repository to understand the specific guidelines and conventions for integrating models into their package.

Unfortunately, I cannot provide direct links or URLs in this text-based conversation, but a quick search for "Ultralytics documentation" or "Ultralytics GitHub repository" should point you in the right direction to find the resources you need. These resources should provide detailed explanations and examples on how to make the necessary changes to make the model Ultralytics-native.

Keep in mind that it's important to focus on adapting the code to work with the Ultralytics package and to remove any dependencies on external libraries or modules like argparse that are not used by the package. This will ensure compatibility and consistency within the Ultralytics framework.

If you have any further questions or need more clarification, feel free to ask. Good luck with making the changes to make your model Ultralytics-native!

**Comment by glenn-jocher** [comment]
@kaushal07wick CI tests is passing but there are a number of precommit issues that need to be resolved. Please take a look at https://results.pre-commit.ci/run/github/535360445/1687506423.VmvqaC4dQqicJUr_xb6jMg

Screenshot 2023-06-24 at 15 49 12

**Comment by glenn-jocher** [comment]
@kaushal07wick for examples of new models added see SAM, RTDETR, YOLO-NAS integrations, and for YOLOv8 models see the v8 directory, i.e.:

https://github.com/ultralytics/ultralytics/tree/main/ultralytics/vit/rtdetr

**Comment by kaushal07wick** [comment]
> @kaushal07wick CI tests is passing but there are a number of precommit issues that need to be resolved. Please take a look at https://results.pre-commit.ci/run/github/535360445/1687506423.VmvqaC4dQqicJUr_xb6jMg
> Screenshot 2023-06-24 at 15 49 12

sure I will look into it.

**Comment by kaushal07wick** [comment]
> Hi @kaushal07wick, to make the model compatible with the Ultralytics package, you wouldn't need to use the argparse library for argument parsing. Instead, you can utilize the configurations and arguments provided by the Ultralytics package itself.
> 
> Ultralytics provides its own way of handling configuration and arguments specifically tailored for their framework. You can refer to the Ultralytics documentation or their GitHub repository to understand the specific guidelines and conventions for integrating models into their package.
> 
> Unfortunately, I cannot provide direct links or URLs in this text-based conversation, but a quick search for "Ultralytics documentation" or "Ultralytics GitHub repository" should point you in the right direction to find the resources you need. These resources should provide detailed explanations and examples on how to make the necessary changes to make the model Ultralytics-native.
> 
> Keep in mind that it's important to focus on adapting the code to work with the Ultralytics package and to remove any dependencies on external libraries or modules like argparse that are not used by the package. This will ensure compatibility and consistency within the Ultralytics framework.
> 
> If you have any further questions or need more clarification, feel free to ask. Good luck with making the changes to make your model Ultralytics-native!

I'll try my best for that.

**Comment by glenn-jocher** [comment]
Hi @kaushal07wick, to make the model compatible with the Ultralytics package, you wouldn't need to use the argparse library for argument parsing. Instead, you can utilize the configurations and arguments provided by the Ultralytics package itself.

Ultralytics provides its own way of handling configuration and arguments specifically tailored for their framework. You can refer to the Ultralytics documentation or their GitHub repository to understand the specific guidelines and conventions for integrating models into their package.

Unfortunately, I cannot provide direct links or URLs in this text-based conversation, but a quick search for "Ultralytics documentation" or "Ultralytics GitHub repository" should point you in the right direction to find the resources you need. These resources should provide detailed explanations and examples on how to make the necessary changes to make the model Ultralytics-native.

Keep in mind that it's important to focus on adapting the code to work with the Ultralytics package and to remove any dependencies on external libraries or modules like argparse that are not used by the package. This will ensure compatibility and consistency within the Ultralytics framework.

If you have any further questions or need more clarification, feel free to ask. Good luck with making the changes to make your model Ultralytics-native!

**Comment by glenn-jocher** [comment]
@ExtReMLapin hi there,

Thank you for reaching out. We're glad to hear that you're excited about the new models being added to our repository.

We understand that it can be frustrating when certain models have specific requirements that can affect your workflow. However, our goal is to provide a wide range of models that can meet different needs and use cases.

Regarding YOLOv8, we suggest taking a look at the YOLOv8 directory in our repository, which includes different variations of the YOLOv8 model. Each variation has its own benefits and characteristics, so you can choose the one that best fits your needs.

If you have any specific questions or concerns about YOLOv8 or any other model in our repository, feel free to let us know. We're always happy to help.

Best, 
The Ultralytics Team

**Comment by kaushal07wick** [comment]
hey @AyushExel @glenn-jocher , I am still little ambiguous in removing argparse dependent variables and making it ultralytics native. Could please provide a small example if possible.

**Comment by glenn-jocher** [comment]
Hey there! 🙌

Sure thing, I'd be happy to provide an example. Instead of using `argparse`, you can directly specify configurations in your code. When integrating with the Ultralytics package, configurations are usually handled through YAML files or directly passing parameters to functions or class constructors.

For instance, if you’re setting up a model for training, instead of parsing arguments, you directly pass the configuration values as parameters:

```python
from ultralytics import Train

# Define your custom configurations
cfg = {
    'model': 'yolov8n.pt',
    'data': 'dataset.yaml',
    'epochs': 100,
    'img_size': 640
}

# Initialize training with configurations
training = Train(**cfg)
training.run()
```

This is a simplified example to illustrate the concept. The actual implementation details might vary based on the specific functionalities you're dealing with. I hope this clarifies things a bit! If you have more questions, feel free to ask. 😊
